base/Makefile

```diff
@@ -104,7 +104,101 @@ else
 endif
 endif
 
-.PHONY: $(BUILDDIR)/build_h.jl.phony $(BUILDDIR)/version_git.jl.phony clean all
+ifneq ($(OS), WINNT)
+# For any USE_SYSTEM_* libraries that will be dynamically loaded by libjulia,
+# put a symlink in the private libdir.
+# System package managers may want to install these links manually,
+# but if not, this will try to create the proper symlink to the right minor version.
+all: symlink_system_libraries
+SYMLINK_SYSTEM_LIBRARIES :=
+
+# if it's a symlink, pick up the symlink target instead, for one level of indirection
+resolve_path = \
+	$1_wd="`pwd`" && \
+	$1_=`readlink -n $${$1} || true` && \
+	if [ -n "$${$1_}" ]; then $1_wd=`dirname "$${$1}"`; $1="$${$1_}"; fi
+## if it's a relative path, make it an absolute path
+resolve_path += && \
+	if [ -z "`echo $${$1} | grep '^/'`" ]; then $1=$${$1_wd}/$${$1}; fi
+ifeq ($(OS), Darwin)
+# try to use the install_name id instead (unless it is an @rpath or such)
+# if it's a relative path, make it an absolute path using the working directory from $1,
+# while in theory incorrect, this has been observed to be a common build mistake for many libraries
+resolve_path += && \
+	$1_=`otool -D $${$1} | tail -n +2 | sed -e 's/^@.*$$//'` && \
+	if [ -n "$${$1_}" ]; then \
+	$1_wd=`dirname "$${$1}"`; $1=$${$1_}; \
+	if [ -z "`echo $${$1} | grep '^/'`" ]; then $1=$${$1_wd}/$${$1}; fi; \
+	fi
+endif
+
+## debug code: `make resolve-path P=<path to test>`
+#resolve_path += && \
+#	echo "$${$1_wd} $${$1}"
+#resolve-path:
+#	$(call resolve_path,P) && \
+#	echo "$$P"
+
+define symlink_system_library
+symlink_$1: $$(build_private_libdir)/$1.$$(SHLIB_EXT)
+$$(build_private_libdir)/$1.$$(SHLIB_EXT):
+	REALPATH=`$$(call spawn,$$(build_depsbindir)/libwhich) -p $$(notdir $$@)` && \
+	$$(call resolve_path,REALPATH) && \
+	[ -e "$$$$REALPATH" ] && \
+	([ ! -e "$$@" ] || rm "$$@") && \
+	echo ln -sf "$$$$REALPATH" "$$@" && \
+	ln -sf "$$$$REALPATH" "$$@"
+ifneq ($2,)
+ifneq ($$(USE_SYSTEM_$2),0)
+SYMLINK_SYSTEM_LIBRARIES += symlink_$1
+endif
+endif
+endef
+
+# the following excludes: libuv.a, libutf8proc.a, suitesparse*
+
+$(eval $(call symlink_system_library,$(LIBMNAME)))
+ifneq ($(USE_SYSTEM_LIBM),0)
+SYMLINK_SYSTEM_LIBRARIES += symlink_$(LIBMNAME)
+else ifneq ($(USE_SYSTEM_OPENLIBM),0)
+SYMLINK_SYSTEM_LIBRARIES += symlink_$(LIBMNAME)
+endif
```

Comment:
I'm confused about this `$(LIBMNAME)` business, it looks like you're calling `symlink_system_library()` on `$(LIBMNAME)` always, but then you add it on to `SYMLINK_SYSTEM_LIBRARIES` which is itself eventually going to `symlink_$(LIBNAME)` eventually?  Aren't we doing work twice here?

Reply:
I'm calling `symlink_system_library` with one argument in this particular case, which skips adding it to `SYMLINK_SYSTEM_LIBRARIES` (so I can do different custom logic here than standard for the rest)